src/main/java/me/modmuss50/optifabric/mod/OptifineSetup.java

```diff
@@ -190,104 +169,114 @@ IMappingProvider createMappings(String from, String to) {
-	}
 
-	//Loom 0.2.7 fallback
-	Path getNewMinecraftDevJar() throws FileNotFoundException {
-		Optional<Path> entrypointResult = getSource(Knot.class.getClassLoader(), "mappings/mappings.tiny");
+		return new ArrayList<>(Arrays.asList(libs));
+	}
 
-		if (entrypointResult.isPresent()) {
-			Path path = entrypointResult.get().getParent();
-			Path minecraftJar = path.resolve(String.format("minecraft-%s-client.jar", OptifineVersion.minecraftVersion)); //Lets hope you are using loom in dev
-			if (Files.exists(minecraftJar)) {
-				return minecraftJar;
+	//Gets the offical minecraft jar
+	private static Path getMinecraftJar() {
+		String givenJar = System.getProperty("optifabric.mc-jar");
+		if (givenJar != null) {
+			File givenJarFile = new File(givenJar);
+
+			if (givenJarFile.exists()) {
+				return givenJarFile.toPath();
+			} else {
+				System.err.println("Supplied Minecraft jar at " + givenJar + " doesn't exist, falling back");
 			}
 		}
 
-		throw new FileNotFoundException("Could not find minecraft jar!");
-	}
+		Path minecraftJar = getLaunchMinecraftJar();
+
+		if (FabricLoader.getInstance().isDevelopmentEnvironment()) {
+			Path officialNames = minecraftJar.resolveSibling(String.format("minecraft-%s-client.jar", OptifineVersion.minecraftVersion));
+
+			if (Files.notExists(officialNames)) {
+				Path parent = minecraftJar.getParent().resolveSibling(String.format("minecraft-%s-client.jar", OptifineVersion.minecraftVersion));
+
+				if (Files.notExists(parent)) {
+					Path alternativeParent = parent.resolveSibling("minecraft-client.jar");
 
-	//Stolen from fabric loader
-	static Optional<Path> findFirstClass(ClassLoader loader, List<String> classNames) {
-		List<String> entrypointFilenames = classNames.stream().map((ep) -> ep.replace('.', '/') + ".class").collect(Collectors.toList());
+					if (Files.notExists(alternativeParent)) {
+						throw new AssertionError("Unable to find Minecraft dev jar! Tried " + officialNames + ", " + parent + " and " + alternativeParent
+								+ "\nPlease supply it explicitly with -Doptifabric.mc-jar");
+					}
+
+					parent = alternativeParent;
+				}
 
-		for (int i = 0; i < entrypointFilenames.size(); i++) {
-			String className = classNames.get(i);
-			String classFilename = entrypointFilenames.get(i);
-			Optional<Path> classSourcePath = getSource(loader, classFilename);
-			if (classSourcePath.isPresent()) {
-				return Optional.of(classSourcePath.get());
+				officialNames = parent;
 			}
+
+			minecraftJar = officialNames;
 		}
 
-		return Optional.empty();
+		return minecraftJar;
 	}
 
-	static Optional<Path> getSource(ClassLoader loader, String filename) {
-		URL url;
-		if ((url = loader.getResource(filename)) != null) {
-			try {
-				URL urlSource = UrlUtil.getSource(filename, url);
-				Path classSourceFile = UrlUtil.asPath(urlSource);
+	private static Path getLaunchMinecraftJar() {
+		try {
+			return (Path) FabricLoader.getInstance().getObjectShare().get("fabric-loader:inputGameJar");
```

Comment:
If you're depending on Loader 0.12 anyway you might as well set 0.12.3 as the minimum then you won't need the catch block (as the object share is always present)

Reply:
If you can make it work on a wider version range that makes things simpler for end users who have existing (old) installs and would like drop-in updates.